### PR TITLE
Feature/widget configurable wrapper

### PR DESCRIPTION
**Pull request by bobdenotter**
Allows themes to set the widget-wrapper and widget-holder HTML. Last functional part of #4444.

### CODE REVIEW

File: src/Asset/Widget/Queue.php [hunk 1]
```diff
@@ -156,23 +156,23 @@ public function countItemsInQueue($location, $type = 'frontend')
...
         }
 
         if ($html !== null) {
-            $html = sprintf(
-                '<div class="widgetholder widgetholder-%s">%s</div>',
-                $location,
-                $html
-            );
+            if (empty($wrappertemplate)) {
+                $wrappertemplate = 'widgetholder.twig';
```
Comment (CarsonF) [inline src/Asset/Widget/Queue.php]: Why not define this in the parameter definition instead of empty string? Same for below as well

Comment (GawainLynch) [inline src/Asset/Widget/Queue.php]: And that should be done at the Twig Extension end… the API parameters should be required.

File: src/Asset/Widget/Queue.php [hunk 2]
```diff
@@ -156,23 +156,23 @@ public function countItemsInQueue($location, $type = 'frontend')
      *
      * @return string|null
      */
-    public function render($location, $type = 'frontend')
+    public function render($location, $type = 'frontend', $wrappertemplate = '', $holdertemplate = '')
     {
```
Comment (GawainLynch) [inline src/Asset/Widget/Queue.php]: Small favour… can we go for camelCase on `$wrapperTemplate` and `$holderTemplate` to keep in-line with PSR-2 and the way we're taking our code standards.

File: src/Asset/Widget/Queue.php (hunk 1 above)
Comment (CarsonF) [inline src/Asset/Widget/Queue.php]: Yes, I would argue there, since it is the entry point, but also here since this could be called from php.

Comment (GawainLynch) [inline src/Asset/Widget/Queue.php]: On that last one, yes… but `$holderTemplate` should be required or default to `null`, and certainly be required in the protected function.

**Comment by bobdenotter** [comment]
Ok, it's currently broken, and I can't figure out why. I have this in my `_aside.twig` on line 7: 

            {{ widgets('aside_top') }}

and I get this error: 

screenshot 2015-12-01 20 32 31

Where it _actually_ breaks, is in the `render()` here: 

```
    protected function addWidgetHolder(WidgetAssetInterface $widget, $holderTemplate)
    {
        $html = $this->render->render(
            $holderTemplate,
            [
                'widget' => $widget,
                'html'   => $widget->isDeferred() ? '' : $this->getHtml($widget),
            ]
        );

        return $html;
    }
```

Debugging tells me `$holderTemplate` is `widgetholder.twig` at that point, and the file exists and is readable. What the hell?!

**Comment by GawainLynch** [comment]
@bobdenotter we've had a look, test is fixed… but @CarsonF think a few things need to change before we deploy API.

1. We should rename the 'Type' use to 'Zone' as that is consistent with how we're doing it in the controllers
2. We should only use one template so we don't sub-render.
  * The exception you're seeing is due to the template shifting around and needed elements from teh widget object that don't exists outside the loop

**Comment by CarsonF** [comment]
@GawainLynch Can we change `Widget` `type` property to `zone` as well?

@bobdenotter I haven't quite figured out how to combine templates yet, so until then carry on.

I am wondering why the twig template files need to be configurable at this level. Isn't the precedence of twig paths enough? I mean that's the whole reason why this folder is called `theme_defaults`.

**Comment by GawainLynch** [comment]
>  Can we change Widget type property to zone as well?

That was what I was asking

**Comment by GawainLynch** [comment]
@bobdenotter happy to offer our service here too… just watching where my foot is in proximity to toes :grin:

**Comment by bobdenotter** [comment]
Green lights! :-)

**Comment by GawainLynch** [comment]
Amanhã… Precisar dormir!

File: src/Asset/Widget/Queue.php [hunk 3]
```diff
@@ -96,7 +96,7 @@ public function process($html)
...
         /** @var WidgetAssetInterface $widget */
         foreach ($this->queue as $widget) {
-            if ($widget->getType() === 'frontend' && $widget->isDeferred()) {
+            if ($widget->getZone() === 'frontend' && $widget->isDeferred()) {
```
Comment (CarsonF) [inline src/Asset/Widget/Queue.php]: Can you use the `Zone::FRONTEND` constant instead of the string?

**Comment by GawainLynch** [comment]
Ninja added the constants… time to :shipit:

**Pull request merged**